README.md
# Test Audit
This cookbook is designed to run as a base policy for a test of the `audit` cookbook managed by Chef Software at https://supermarket.chef.io/cookbooks/audit.

## Platforms
The Test Audit cookbook works for the following platforms (tested December 2019):
- Windows Server 2012 R2
- Windows Server 2016
- RedHat Enterprise Linux 8
- Ubuntu 18.04

## Auditing
It wraps `audit` and pushes out to the Chef Infra server, which should in turn push out to the Chef Automate server for reporting.

It uses these audit profiles:
- admin/linux-baseline
- admin/thosey-windows-baseline

## Wrapped Cookbooks
The Test Audit cookbook wraps the following cookbooks:
- [audit](https://supermarket.chef.io/cookbooks/audit) cookbook
  - Provides auditing via compliance profiles (InSpec)
- [chef-client](https://supermarket.chef.io/cookbooks/chef-client) cookbook
  - Allows for scheduling of tasks/cron jobs for Chef Client runs
- [line](https://supermarket.chef.io/cookbooks/line) cookbook
  - Makes searching and replace of lines in configuration files very easy

## Remediations
This cookbook provides remediations to bring boxes of the above platforms into compliance.

Check the `windows.rb` and `ubuntu.rb` recipes for the remediations included.

## Tests
Contained within is a kitchen.yml file that specifically fetches the InSpec profiles and uses those to test. It also has a `default_tests.rb` file that describes some of the more basic Windows and Ubuntu changes we expect to see at the end of a run.


CHANGELOG.md
# test-audit CHANGELOG

This file is used to list changes made in each version of the test-audit cookbook.

## 0.3.17

- Cleanup of recipes -- relying on chef-client cookbook without direct calls
- Removed default from attributes for `chef-client` cookbook interval

## 0.3.16

- Removed recipe calls in `default` recipe
- Modified `Policyfile.rb` to run_list `audit` and `chef-client`

## 0.3.15

- Removed `remediate.rb` resource and merged code into main Linux recipe

## 0.3.14

- Changed `login.defs` file mode to be 0444 on whatever version of Linux

## 0.3.13

- Changed `file` to `cookbook_file` for `*-login.defs` files

## 0.3.12

- Moved `replace_or_add` lines at start to be individually managed files - `rhel-login.defs` and `ubuntu-login.defs`
- Added some variables/ifs for mode and file based on platform
- Updated `readme.md`

## 0.3.11

- Completely removing `/usr/libexec/openssh/ssh-keysign`

## 0.3.10

- Changed `os10` action of the `remediate` resource to use `append_if_no_line`

## 0.3.9

- Reset contents of `/usr/libexec/openssh/ssh-keysign`

## 0.3.8

- Added centos-8 to `kitchen.yml` for better testing/integration
- Made `auditd` package for Ubuntu only
- Added `line` cookbook dependency
- Changed `haveged` package to be Ubuntu-only
- Added sysctl-02
- Tweaked the `sed` commands to use the line cookbook's `replace_or_add` resource

## 0.3.7

- Added rhel8 to `kitchen.yml` for quicker testing/iteration
- Condensed both Ubuntu and RHEL recipes into a single `linux.rb` recipe.

## 0.3.6

- Added os-05b remediation
- Skipping os-06 for now since I don't know repercussions of clearing that file right now
- Changed `remediate` custom resource action from `ubuntu_os10` to just `os10` as it works for RHEL too

## 0.3.5

- Added os-05 remediations
- Added remediations (recipes/rhel8.rb) recipe for RedHat 8
- Called remediations recipe in default.rb
- Added execute lines that checks for existence of `ENV_PATH` and `ENV_SUPATH` and add if not exist
- Changed from using `=` to `\t` for defining `ENV_PATH` and `ENV_SUPATH`

## 0.3.4

- Added support for RedHat
  - Added `supports` line
  - Added to attributes/default.rb `case` statement

## 0.3.3

- Updated `default_test.rb` and `kitchen.yml` to focus on testing for resources and InSpec verification
- Updated version
- Finalizing before peer review

## 0.3.2

- Updated havaged to notify the shell execution if installed

## 0.3.1

- Changed all the `execute` resources calling sysctl into `sysctl` resources for Ubuntu

## 0.3.0

- Added `chef-client` cookbook dependency so we can have auto runs.

## 0.2.3

- Adding final MINOR remediations for some registry keys in Win

## 0.2.2

- Re-added Set Access Credential Manager as a trusted caller to No One
- Created wrapper policy

## 0.2.1

- Remediated another user right
- Removed network access right (commented) as it blocked any remote access to the server

## 0.2.0

- Created `windows_userright.rb` resource that removes all users from right and adds specified users.

## 0.1.24

- Changed to Grant-UserRight for the powershell module

## 0.1.23

- Reviewed how I was handliong privilege revocation

## 0.1.22

- Changed `NullSessionShares` to string_multi.

## 0.1.21

- Added the remainign registry key remediations (I think)

## 0.1.20

- Added some registry key remediations.

## 0.1.19

- Testing out a Windows Powershell module for setting/revoking rights

## 0.1.18

- Began windows remediations
- Added ntrights.exe to the files

## 0.1.17

- Remediating max_log_file_action for auditd

## 0.1.16

- Adding remediation for `package-08` to install auditd

## 0.1.15

- Screwed up one of the remediations. Fixed.

## 0.1.14

- Added remediations for os-05, os-08 and os-10 for ubuntu
- Added remediations for sysctl-05 to 10, 13-18, 20-28, 30, 31b
- Tweaks to Resource

## 0.1.13

- Cleaned up attributes spacing
- Corrected from `node['os']` to `node['platform']`

## 0.1.12

- Tweaked the include_recipe lines to vary based on OS

## 0.1.11

- Added recipes for Win and Ubuntu with case to pick between the two
- Remediation for os-08 (Entropy increase)

## 0.1.10

- Fixing an issue with profiles not being proper hashes

## 0.1.9

- Clarified description
- Added dependency for `audit`
- Added OS `supports` lines

## 0.1.8

- Added audit cookbook init in `recipes/default.rb`

## 0.1.7

- Changed the way the default profile is specified in the attribs
- Changed back to a switch/when case statement

## 0.1.6

- Tweaked attributes case to a pair of if statements

## 0.1.5

- Added attributes file
- Moved everything to attributes

## 0.1.4

- Changed to linux-baseline and windows-baseline for the profiles to start
- Added `data_collector` and `profiles`

## 0.1.3

- Changed required Chef version to 14 since we don't require anything from 15

## 0.1.2

- Corrected use of `default` to `node`

## 0.1.1

- Added wrapped policy `audit`
- Set to send to Automate via Chef Infra Server
- Added fork for Windows and Ubuntu
  - Runs cis-windows2016rtm-release1607-level1-memberserver for Win
  - Runs cis-ubuntu18.04lts-level1-server for Ubuntu

## 0.1.0

Initial release.

- Base generated cookbook.



recipes/linux.rb
# os-05 remediation for login.defs file and os-05b RedHat-specific remediations for login.defs
login_defs_file = if platform?('ubuntu')
                    'ubuntu-login.defs'
                  else
                    'rhel-login.defs'
                  end

cookbook_file '/etc/login.defs' do
  source login_defs_file
  mode '0444'
  owner 'root'
  group 'root'
end

# os-06 clearing /usr/libexec/openssh/ssh-keysign
file '/usr/libexec/openssh/ssh-keysign' do
  action :delete
end

# Add lines to /etc/modprobe.d/dev-sec.conf if they don't exist (os-10)
entries = [
    'install cramfs /bin/true',
    'install freevxfs /bin/true',
    'install jffs2 /bin/true',
    'install hfs /bin/true',
    'install hfsplus /bin/true',
    'install squashfs /bin/true',
    'install udf /bin/true',
    'install vfat /bin/true',
  ]
entries.each do |entry|
  append_if_no_line entry do
    path '/etc/modprobe.d/dev-sec.conf'
    line entry
  end
end

# Install auditd to remediate package-08
package 'auditd' do
  only_if { platform?('ubuntu') }
end

# Update max_log_file_action
replace_or_add 'Audit max_log_file_action' do
  path '/etc/audit/auditd.conf'
  pattern 'max_log_file_action*'
  line 'max_log_file_action = keep_logs'
end

replace_or_add 'Audit log_format' do
  path '/etc/audit/auditd.conf'
  pattern 'log_format*'
  line 'log_format = raw'
end

# Install havaged package to increase entropy (os-08)
package 'haveged' do
  notifies :run, 'execute[update-rc.d haveged defaults]', :immediate
  only_if { platform?('ubuntu') }
end

execute 'update-rc.d haveged defaults' do
  action :nothing
end

# Apply changes to various kernel parameters - see sysctl-02, 05-10, 13-18,
# 20-28, 30, 31b

# sysctl-05-10
sysctl 'net.ipv4.icmp_ratelimit' do
  value 100
end

sysctl 'net.ipv4.icmp_ratemask' do
  value 88089
end

sysctl 'net.ipv4.conf.all.arp_announce' do
  value 2
end

# sysctl31b
sysctl 'kernel.core_pattern' do
  value '/^\\/.*/'
end

# sysctl-02, sysctl-13 to 18, sysctl-20 to 28, # sysctl-30
# Sets all these values to 0
zero_set = [
  'net.ipv4.tcp_timestamps',
  'net.ipv4.conf.default.accept_source_route',
  'net.ipv4.conf.default.accept_redirects',
  'net.ipv4.conf.all.accept_redirects',
  'net.ipv4.conf.all.secure_redirects',
  'net.ipv4.conf.default.secure_redirects',
  'net.ipv4.conf.default.send_redirects',
  'net.ipv4.conf.all.send_redirects',
  'net.ipv6.conf.default.accept_redirects',
  'net.ipv6.conf.all.accept_redirects',
  'net.ipv6.conf.default.router_solicitations',
  'net.ipv6.conf.default.accept_ra_rtr_pref',
  'net.ipv6.conf.default.accept_ra_pinfo',
  'net.ipv6.conf.default.accept_ra_defrtr',
  'net.ipv6.conf.all.accept_ra',
  'net.ipv6.conf.default.accept_ra',
  'net.ipv6.conf.default.autoconf',
  'net.ipv6.conf.default.dad_transmits',
  'kernel.sysrq',
]

zero_set.each do |this_set|
  sysctl this_set do
    value 0
  end
end

# Sets all these values to 1

one_set = [
  'net.ipv4.conf.default.rp_filter',
  'net.ipv4.conf.all.arp_ignore',
  'net.ipv4.tcp_rfc1337',
  'net.ipv4.conf.all.log_martians',
  'net.ipv4.conf.default.log_martians',
  'net.ipv6.conf.all.disable_ipv6',
  'net.ipv6.conf.default.max_addresses',
]

one_set.each do |this_set|
  sysctl this_set do
    value 1
  end
end
